modules/uploads/scripts/complete.ts

```diff
@@ -0,0 +1,135 @@
+import { RuntimeError, ScriptContext } from "../_gen/scripts/complete.ts";
+import { keyExists, getMultipartUploadParts, completeMultipartUpload } from "../utils/bucket.ts";
+import { getS3EnvConfig } from "../utils/env.ts";
+import { getKey } from "../utils/types.ts";
+import { prismaToOutput } from "../utils/types.ts";
+import { Upload } from "../utils/types.ts";
+
+export interface Request {
+	uploadId: string;
+}
+
+export interface Response {
+	upload: Upload;
+}
+
+export async function run(
+	ctx: ScriptContext,
+	req: Request,
+): Promise<Response> {
+	const s3 = getS3EnvConfig(ctx.userConfig.s3);
+	if (!s3) throw new RuntimeError("s3_not_configured");
+
+	const newUpload = await ctx.db.$transaction(async (db) => {
+		// Find the upload by ID
+		const upload = await db.upload.findFirst({
+			where: {
+				id: req.uploadId,
+			},
+			select: {
+				id: true,
+				userId: true,
+				bucket: true,
+				contentLength: true,
+				files: true,
+				createdAt: true,
+				updatedAt: true,
+				completedAt: true,
+			},
+		});
+
+		// Error if the upload wasn't prepared
+		if (!upload) {
+			throw new RuntimeError(
+				"upload_not_found",
+				{
+					meta: { uploadId: req.uploadId },
+				},
+			);
+		}
+
+		// Check with S3 to see if the files were uploaded
+		const fileExistencePromises = upload.files.map(
+			async file => {
+				// If the file was uploaded in parts, complete the multipart upload
+				if (file.multipartUploadId) {
+					try {
+						const parts = await getMultipartUploadParts(
+							s3,
+							getKey(upload.id, file.path),
+							file.multipartUploadId,
+						);
+						if (parts.length === 0) return false;
+
+						await completeMultipartUpload(
+							s3,
+							getKey(upload.id, file.path),
+							file.multipartUploadId,
+							parts,
+						);
+
+
+					} catch (e) {
+						throw new RuntimeError(
+							"multipart_upload_completion_fail",
+							{ cause: e },
+						)
+					}
+				}
+
+				// Check if the file exists
+				return await keyExists(s3, getKey(upload.id, file.path))
```

Comment:
Can you read up on if this is redundant for multipart uploads? I think this doesn't do anything if `completeMultipartUpload` passes.

Reply:
Reading further into the documentation, I believe the SDK handles errors relating to the objects not being created.